kasa/iot/iottimezone.py

```diff
@@ -23,26 +23,30 @@ async def get_timezone(index: int) -> tzinfo:
     return await _CachedZoneInfo.get_cached_zone_info(name)
 
 
-async def get_timezone_index(name: str) -> int:
+async def get_timezone_index(tzone: tzinfo) -> int:
     """Return the iot firmware index for a valid IANA timezone key."""
-    rev = {val: key for key, val in TIMEZONE_INDEX.items()}
-    if name in rev:
-        return rev[name]
+    if isinstance(tzone, ZoneInfo):
+        name = tzone.key
+        rev = {val: key for key, val in TIMEZONE_INDEX.items()}
+        if name in rev:
+            return rev[name]
 
-    # Try to find a supported timezone matching dst true/false
-    zone = await _CachedZoneInfo.get_cached_zone_info(name)
-    now = datetime.now()
-    winter = datetime(now.year, 1, 1, 12)
-    summer = datetime(now.year, 7, 1, 12)
     for i in range(110):
-        configured_zone = await get_timezone(i)
-        if zone.utcoffset(winter) == configured_zone.utcoffset(
-            winter
-        ) and zone.utcoffset(summer) == configured_zone.utcoffset(summer):
+        if _is_same_timezone(tzone, await get_timezone(i)):
             return i
     raise ValueError("Device does not support timezone %s", name)
 
 
+def _is_same_timezone(tzone1: tzinfo, tzone2: tzinfo) -> bool:
+    now = datetime.now()
```

Comment:
Will do re: the docstring.

Basically the answer to this question relates to your observation above and getting the `Africa/Windhoek` timezone. Because the firmware doesn't use the IANA standard list and it only has 109 timezones setup this function looks for a timezone that matches the IANA one provided by going through the list and checking the utc offset is the same for an entire year.

When you do the `sync` without providing a `--timezone` parameter it takes the local timezone returned by `datetime.now().astimezone().tzinfo` which in your case is probably coming back as "CEST" or "CEDT", however this is from the `timezone` lib and it's not dst aware so it just gives a fixed UTC offset which doesn't match with CET but instead matches with `Africa/Windhoek`.

This might make it clearer:
```py
>>> from kasa.iot.iottimezone import _is_same_timezone
>>> from zoneinfo import ZoneInfo
>>> mytz = datetime.now().astimezone().tzinfo
>>> mytz
datetime.timezone(datetime.timedelta(seconds=3600), 'BST')
>>> _is_same_timezone(mytz, ZoneInfo("Europe/London"))
False
>>> _is_same_timezone(mytz, ZoneInfo("Etc/GMT-1")) # Etc/GMT-1 is UTC+1:00, no dst
True
```

So thinking about this some more we should make the `set_time` api handle setting the time without updating the timezone (using whatever timezone is currently configured). This will mean excluding the `--timezone`` parameter will not fall into this problem.